File: FP/Lab 6/action.py
class Action:
  def __init__(self, repo):
    self.__repo = repo

  def repo(self):
    return self.__repo
    
  def apply(self):
    pass
  
  def roll_back(self):
    pass
  
class AddAction(Action):
  def __init__(self, repo, val):
    Action.__init__(self, repo)
    self.__val = val

  def apply(self):
    self.repo().add(self.__val)

  def roll_back(self):
    self.repo().remove(self.__val)

class RemoveAction(Action):
  def __init__(self, repo, val):
    Action.__init__(self, repo)
    self.__val = val

  def apply(self):
    self.repo().remove(self.__val)

  def roll_back(self):
    self.repo().add(self.__val)

class UpdateAction(Action):
  def __init__(self, repo, old_val, new_val):
    Action.__init__(self, repo)
    self.__old_val = old_val
    self.__new_val = new_val

  def apply(self):
    self.repo().update(self.__new_val)

  def roll_back(self):
    self.repo().update(self.__old_val)

File: Y1S2/OOP/Seminar5/honing-your-loom/Action.cc
#include "Action.hh"

Action::Action(Repo& repo) : _repo{repo} { }

AddAction::AddAction(Repo& repo, Task addedTask) : Action{repo}, _addedTask{addedTask} { }

void AddAction::undo() {
  _repo.remove(_addedTask.title());
}

void AddAction::redo() {
  _repo.add(_addedTask);
}

RemoveAction::RemoveAction(Repo& repo, Task removedTask) : Action{repo}, _removedTask{removedTask} { }

void RemoveAction::undo() {
  _repo.add(_removedTask);
}

void RemoveAction::redo() {
  _repo.remove(_removedTask.title());
}

UpdateAction::UpdateAction(Repo& repo, Task oldTask, Task newTask) : Action{repo}, _oldTask{oldTask}, _newTask{newTask} { }

void UpdateAction::undo() {
  _repo.update(_oldTask);
}

void UpdateAction::redo() {
  _repo.update(_newTask);
}